src/async_client.rs

```diff
@@ -540,28 +403,14 @@ impl AsyncClient {
             // After the room has been created and state/timeline events accounted for we use the room_id of the newly created
             // room to add any presence events that relate to a user in the current room. This is not super
             // efficient but we need a room_id so we would loop through now or later.
-            for presence in &response.presence.events {
-                let mut client = self.base_client.write().await;
-                if let EventResult::Ok(e) = presence {
-                    client.receive_presence_event(&room_id_string, e);
-                }
+            for presence in &mut response.presence.events {
+                {
+                    if let EventResult::Ok(e) = presence {
+                        client.receive_presence_event(&room_id_string, e).await;
 
-                let callbacks = {
-                    let mut cb_futures = self.presence_callbacks.lock().unwrap();
-                    let event = if !cb_futures.is_empty() {
-                        Arc::new(presence.clone())
-                    } else {
-                        continue;
-                    };
-                    let mut callbacks = Vec::new();
-                    for cb in &mut cb_futures.iter_mut() {
-                        callbacks.push(cb(matrix_room.clone(), Arc::clone(&event)));
+                        // TODO should we determine if any room state has changed before calling
```

Comment:
While it is possible to send out events that don't do anything and the server will happily forward them to us, this isn't really something that's happening and we can probably assume that if we get a new event from the server, the client will want to know about it.